src/lib/QueryRendererHelper.ts

```diff
@@ -32,10 +31,9 @@ export function explainResults(source: string): string {
     const tasksBlockQuery = new Query({ source });
 
     if (!tasksBlockQuery.ignoreGlobalQuery) {
-        const globalQuery: IQuery = new Query(getGlobalQuerySource());
-
-        if (globalQuery.source.trim() !== '') {
-            result += `Explanation of the global query:\n\n${globalQuery.explainQuery()}\n`;
+        const globalQuery = GlobalQuery.getInstance();
```

Comment:
By using `getInstance()` here and in `explainResults()`, it ties all uses of this code to the global instance, which loses the purpose of this refactoring.

I think this is what must have been unclear from my explanation earlier ... and which in turn is why you were unhappy with the number of calls to `getInstance()`...

Instead, try changing this function so that it takes a GlobalQuery **parameter**.

Now, the code can be supplied a **GlobalQuery** object with any data in it, without needing to change the settings in order to test it...

```diff
Index: src/lib/QueryRendererHelper.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/lib/QueryRendererHelper.ts b/src/lib/QueryRendererHelper.ts
--- a/src/lib/QueryRendererHelper.ts	(revision a697c2273d1d27f3087aec59b8d7b9f04a555cf7)
+++ b/src/lib/QueryRendererHelper.ts	(date 1692127274325)
@@ -1,6 +1,6 @@
-import { GlobalFilter } from '../Config/GlobalFilter';
-import { Query } from '../Query/Query';
-import { GlobalQuery } from '../Config/GlobalQuery';
+import {GlobalFilter} from '../Config/GlobalFilter';
+import {Query} from '../Query/Query';
+import type {GlobalQuery} from '../Config/GlobalQuery';
 
 /**
  * @summary
@@ -20,8 +20,9 @@
  *
  * @param {string} source The source of the task block to explain
  * @returns {string}
+ * @param globalQuery
  */
-export function explainResults(source: string): string {
+export function explainResults(source: string, globalQuery: GlobalQuery): string {
     let result = '';
 
     if (!GlobalFilter.isEmpty()) {
@@ -31,7 +32,6 @@
     const tasksBlockQuery = new Query({ source });
 
     if (!tasksBlockQuery.ignoreGlobalQuery) {
-        const globalQuery = GlobalQuery.getInstance();
         if (!globalQuery.isEmpty()) {
             result += `Explanation of the global query:\n\n${globalQuery.query().explainQuery()}\n`;
         }
@@ -49,14 +49,14 @@
  *
  * @param {string} source The query source from the task block
  * @returns {Query} The query to execute
+ * @param globalQuery
  */
-export function getQueryForQueryRenderer(source: string): Query {
+export function getQueryForQueryRenderer(source: string, globalQuery: GlobalQuery): Query {
     const tasksBlockQuery = new Query({ source });
 
     if (tasksBlockQuery.ignoreGlobalQuery) {
         return tasksBlockQuery;
     }
 
-    const globalQuery = GlobalQuery.getInstance().query();
-    return globalQuery.append(tasksBlockQuery);
+    return globalQuery.query().append(tasksBlockQuery);
 }

```